proposals/nnnn-strict-memory-safety.md

```diff
@@ -0,0 +1,437 @@
+
+### `@unsafe` attribute
+
+The `@unsafe` attribute can be applied to any declaration to indicate that use of that declaration can undermine memory safety. Any use of a declaration marked `@unsafe` will result in a warning. The closest analogue in the language today is `@available(*, deprecated)`, which has effectively no impact on the type system, yet any use of a deprecated declaration results in a warning.
+
+When a type is marked `@unsafe`, a declaration that uses that type in its interface is implicitly `@unsafe`. For example, consider a program containing three separate modules:
+
+```swift
+// Module A
+@unsafe
+public struct DataWrapper {
+  var buffer: UnsafeBufferPointer<UInt8>
+  
+  public func checksum() -> Int32 { ... }
+}
+
+// Module B
+import A
+
+public struct MyType {
+  public var wrapper: DataWrapper
+}
+
+// Module C
+import A
+import B
+
+extension MyType {
+  public func checksum() -> Int32 {}
+    return wrapper.checksum()
+  }
+}
+```
+
+Module `A` defines a type, `DataWrapper`, that is `@unsafe`. It can be compiled with or without strict safety checking enabled, and is fine either way.
+
+Module `B` uses the `DataWrapper` type. If compiled without strict safety checking, there will be no diagnostics about memory safety. If compiled with strict safety checking, there will be a diagnostic about `wrapper` using an `@unsafe` type (`DataWrapper`) in its interface. This diagnostic can be ignored.
+
+In module `C` enables strict memory safety, the use of `MyType` is considered safe (since it was not marked `@unsafe` and doesn't involve unsafe types in its interface). However, the access to `wrapper` will result in a diagnostic, because the type of `wrapper` involves an `@unsafe` type. 
+
+### Encapsulating unsafe behavior
+
+When a declaration is marked `@unsafe`, it is free to use any other unsafe declarations as part of its interface or implementation. In the example from the previously section, both `wrapper` and C's `checksum` can be marked as `@unsafe` to suppress diagnostics by explicitly propagating unsafety to their clients:
+
+```swift
+// Module B
+import A
+
+public struct MyType {
+  @unsafe public var wrapper: DataWrapper
+}
+
+// Module C
+import A
+import B
+
+extension MyType {
+  @unsafe public func checksum() -> Int32 {}
+    return wrapper.checksum()
+  }
+}
+```
+
+However, these two APIs diff in the severity of the problem they post for memory safety: `wrapper` has an unsafe type in its interface, so any use of `wrapper` is fundamentally unsafe. However, `MyType` itself can encapsulate the memory-unsafe behavior within a safe API., The `MyType.checksum` operation is actually safe for clients to use. Marking it as `@unsafe` is therefore undesirable, because it (incorrectly) forces clients to treat this as an unsafe API, causing unnecessary extra work and deluting the value of correctly-identified unsafe APIs.
```

Comment:
```suggestion
However, these two APIs differ in the severity of the problem they post for memory safety: `wrapper` has an unsafe type in its interface, so any use of `wrapper` is fundamentally unsafe. However, `MyType` itself can encapsulate the memory-unsafe behavior within a safe API., The `MyType.checksum` operation is actually safe for clients to use. Marking it as `@unsafe` is therefore undesirable, because it (incorrectly) forces clients to treat this as an unsafe API, causing unnecessary extra work and deluting the value of correctly-identified unsafe APIs.
```